presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecutorFactory.java

```diff
@@ -991,6 +1008,52 @@ private Tuple2<MutablePartitionId, T> doComputeNext()
         }
 
+        private void collectMetrics(TaskInfo taskInfo)
+        {
+            int taskId = taskInfo.getTaskId().getId();
+            int stageId = taskInfo.getTaskId().getStageExecutionId().getStageId().getId();
+            List<Map<String, String>> newStatsList = new ArrayList<>();
+            for (PipelineStats pipelineStats : taskInfo.getStats().getPipelines()) {
+                for (OperatorStats operatorStats : pipelineStats.getOperatorSummaries()) {
+                    if (operatorStats.getOperatorType().equals("NativeExecutionOperator")) {
+                        NativeExecutionInfo nativeExecutionInfo = (NativeExecutionInfo) operatorStats.getInfo();
+                        for (TaskStats taskStats : nativeExecutionInfo.getTaskStats()) {
+                            RuntimeStats runtimeStat = taskStats.getRuntimeStats();
+                            Map<String, String> newStatMap = new HashMap<>();
+                            for (Map.Entry<String, RuntimeMetric> entry : runtimeStat.getMetrics().entrySet()) {
+                                String key = entry.getKey();
+                                RuntimeMetric metric = entry.getValue();
+                                if (metric.getCount() == 0) {
+                                    continue;
+                                }
+                                String metricSumKey = buildGenericMetricKey(taskId, stageId, key, "sum");
+                                newStatMap.put(metricSumKey, String.valueOf(metric.getSum()));
+                                String metricCountKey = buildGenericMetricKey(taskId, stageId, key, "count");
+                                newStatMap.put(metricCountKey, String.valueOf(metric.getCount()));
+                                String metricMinKey = buildGenericMetricKey(taskId, stageId, key, "min");
+                                newStatMap.put(metricMinKey, String.valueOf(metric.getMin()));
+                                String metricMaxKey = buildGenericMetricKey(taskId, stageId, key, "max");
+                                newStatMap.put(metricMaxKey, String.valueOf(metric.getMax()));
+                                String metricUnitKey = buildGenericMetricKey(taskId, stageId, key, "unit");
+                                newStatMap.put(metricUnitKey, String.valueOf(metric.getUnit()));
+                            }
+                            if (!newStatMap.isEmpty()) {
+                                newStatsList.add(newStatMap);
+                            }
+                        }
+                    }
+                }
+            }
+            if (!newStatsList.isEmpty()) {
+                genericShuffleStatsCollector.add(newStatsList);
+            }
+        }
+
+        private String buildGenericMetricKey(int taskId, int stageId, String key, String metricType)
+        {
+            return String.format("%1$s|%2$s|%3$s|%4$s", taskId, stageId, key, metricType);
```

Comment:
Why are we using this format? (`|` separated values).
How would it work if we add a new field to the key, for example stage attempt number?

Reply:
These fields are Presto RuntimeStats fields. RuntimesStats is not available in Spark project since we don't want to create any dependency between both projects.
 
I don't think this will be the right place for stage attempt since this is taskMetric.  

I think min, max, count, sum and unit  should be enough to describe any datapoint till we create a common metrics library for both project.

Reply:
My question is more about how did we finalize on this string pattern (| separated values). Where and how is this pattern used, what is the implication?